_posts/2023-03-27-group-equivariant-cnn-2/group-equivariant-cnn-2.Rmd

```diff
@@ -0,0 +1,374 @@
+elems <- C_4$elements()
+elems
+```
+
+    torch_tensor
+     0.0000
+     1.5708
+     3.1416
+     4.7124
+    [ CPUFloatType{4} ]
+
+Elements are straightforwardly represented by their respective rotation angles: $0$, $\frac{\pi}{2}$, $\pi$, and $\frac{3 \pi}{2}$.
+
+Groups are aware of the identity, and know how to construct an element's inverse:
+
+```{r}
+C_4$identity
+
+g1 <- elems[2]
+C_4$inverse(g1)
+```
+
+    torch_tensor
+     0
+    [ CPUFloatType{1} ]
+
+    torch_tensor
+    4.71239
+    [ CPUFloatType{} ]
+
+Here, what we care about most is the group elements' *action*. Implementation-wise, we need to distinguish between them acting on each other, and their action on the vector space $\mathbb{R}^2$, where our input images live. The former part is the easy one: It may simply be implemented by adding angles. In fact, this is what `gcnn` does when we ask it to let `g1` act on `g2`:
+
+```{r}
+g2 <- elems[3]
+
+# in C_4$left_action_on_H(), H stands for the symmetry group
+C_4$left_action_on_H(torch_tensor(g1)$unsqueeze(1), torch_tensor(g2)$unsqueeze(1))
+```
+
+    torch_tensor
+     4.7124
+    [ CPUFloatType{1,1} ]
+
+What's with the `unsqueeze()`s? Since $C_4$'s ultimate *raison d'être* is to be part of a neural network, `left_action_on_H()` works with batches of elements, not scalar tensors.
+
+Things are a bit less straightforward where the group action on $\mathbb{R}^2$ is concerned. Here, we need the concept of a [group representation](https://en.wikipedia.org/wiki/Group_representation). This is an involved topic, which we won't go into here. In our current context, it works about like this: We have an input signal, a tensor we'd like to operate on in some way. (That "some way" will be convolution, as we'll see soon.) To render that operation group-equivariant, we first have the representation apply the *inverse* group action to the input. That accomplished, we go on with the operation as though nothing had happened.
+
+To give a real-life -- sort of -- example, say the operation is a measurement. We'd like to measure the height of the person in the black running outfit and green shoes, standing at the foot of the trail, ready to run up the climb. One option we have is to take the measurement, then let the person run up. Our measurement will be as valid up the mountain as it was down here. Alternatively, we might be polite and not make them wait. Once they're up there, we ask them to come down, and when they're back, we measure their height. The result is the same: Body height is equivariant (more than that: invariant, even) to the action of running up or down. (Of course, height is a pretty dull measure. But something more interesting, such as heart rate, would not have worked so well in this example.)
+
+Returning to things implementation, it turns out that group actions are encoded as matrices. There is one matrix for each group element. For $C_4$, the so-called *regular* representation is a rotation matrix:
```

Comment:
Do you mean "things" implementation?

Reply:
@batpigandme can you explain?
I guess I do, but it's not clear? And quotes ...idk? Should I write sth different (less colloquial) entirely?

Reply:
no ... more like "after this excursion, get back to the code" ...
then probably "things code" isn't better ...

"this implementation" is a bit different in intent, but correct ... since the following sentence is exactly about  _this_ implementation ...  but maybe I'll write "the implementation", that one's totally neutral